File: bindings/go/backend/metric.go
package backend

import "context"

type MetricKind string

const (
	MetricCounter   MetricKind = "counter"
	MetricGauge     MetricKind = "gauge"
	MetricHistogram MetricKind = "histogram"
)

type MetricSample struct {
	Name        string
	Kind        MetricKind
	BackendName string
	Value       float64
	Unit        string
	Labels      map[string]JSON
}

type MetricInstrumentDefinition struct {
	Name        string
	Kind        MetricKind
	BackendName string
	Unit        string
	Labels      []FieldDescriptor
}

type MetricInstrumentRegistration struct {
	RegisteredNew bool
	Definition    MetricInstrumentDefinition
}

type MetricSink interface {
	// RegisterDefinition is idempotent. Re-registering an identical definition is
	// a no-op; incompatible redefinition should return an error.
	RegisterDefinition(ctx context.Context, backend Backend, definition MetricInstrumentDefinition) (MetricInstrumentRegistration, error)

	RegisterDefinitionTx(ctx context.Context, tx Transaction, definition MetricInstrumentDefinition) (MetricInstrumentRegistration, error)

	// RecordMetricTx stages the sample in the caller's OCC transaction. Commit
	// makes it visible to exporters; rollback drops it. Implementations should
	// validate labels against the registered metric definition.
	RecordMetric(ctx context.Context, backend Backend, sample MetricSample) error

	RecordMetricTx(ctx context.Context, tx Transaction, sample MetricSample) error
}

File: bindings/rust/metric.rs
use std::collections::BTreeMap;

use crate::backend::{Backend, BackendResult, FieldDescriptor};
use crate::json::Json;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MetricKind {
    Counter,
    Gauge,
    Histogram,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MetricSample {
    pub name: String,
    pub kind: MetricKind,
    pub backend_name: String,
    pub value: f64,
    pub unit: String,
    pub labels: BTreeMap<String, Json>,
}

#[derive(Debug, Clone)]
pub struct MetricDefinition {
    pub name: String,
    pub kind: MetricKind,
    pub backend_name: String,
    pub unit: String,
    pub labels: Vec<FieldDescriptor>,
}

#[derive(Debug, Clone)]
pub struct MetricDefinitionRegistration {
    pub registered_new: bool,
    pub definition: MetricDefinition,
}

pub trait MetricSink<B: Backend> {
    fn register_definition(
        &self,
        backend: &B,
        definition: &MetricDefinition,
    ) -> BackendResult<MetricDefinitionRegistration>;

    fn register_definition_tx(
        &self,
        tx: &mut B::Tx,
        definition: &MetricDefinition,
    ) -> BackendResult<MetricDefinitionRegistration>;

    /// Transactional records are staged in the caller's OCC transaction. Commit
    /// makes the sample visible to exporters; rollback drops it. Implementations
    /// should validate labels against the registered metric definition.
    fn record_metric(&self, backend: &B, sample: &MetricSample) -> BackendResult<()>;

    fn record_metric_tx(&self, tx: &mut B::Tx, sample: &MetricSample) -> BackendResult<()>;
}
